src/proto/media/partition.rs

```diff
@@ -0,0 +1,202 @@
+//! Partition information protocol.
+
+use crate::proto::Protocol;
+use crate::{unsafe_guid, Char16, Guid};
+
+newtype_enum! {
+    /// MBR OS type.
+    ///
+    /// Only two values are defined in the UEFI specification (section
+    /// 5.2.2), other values are used by legacy operating systems.
+    pub enum MbrOsType: u8 => {
+        /// A fake partition covering the entire disk.
+        GPT_PROTECTIVE = 0xee,
+
+        /// UEFI system partition.
+        UEFI_SYSTEM_PARTITION = 0xef,
+    }
+}
+
+/// Legacy MBR partition record.
+///
+/// UEFI specification: table 5-2.
+#[repr(C)]
+#[repr(packed)]
+#[derive(Clone, Copy, Debug)]
+pub struct MbrPartitionRecord {
+    boot_indicator: u8,
+    starting_chs: [u8; 3],
+    os_type: MbrOsType,
+    ending_chs: [u8; 3],
+    starting_lba: u32,
+    size_in_lba: u32,
+}
+
+impl MbrPartitionRecord {
+    /// True if the partition is a bootable legacy partition.
+    pub fn is_bootable(&self) -> bool {
+        self.boot_indicator == 0x80
+    }
+
+    /// Start of the partition in CHS address format.
+    pub fn starting_chs(&self) -> [u8; 3] {
+        self.starting_chs
+    }
+
+    /// End of the partition in CHS address format.
+    pub fn ending_chs(&self) -> [u8; 3] {
+        self.ending_chs
+    }
+
+    /// Type of partition.
+    pub fn os_type(&self) -> MbrOsType {
+        self.os_type
+    }
+
+    /// Starting LBA of the partition on the disk.
+    pub fn starting_lba(&self) -> u32 {
+        self.starting_lba
+    }
+
+    /// Size of the partition in LBA units of logical blocks.
+    pub fn size_in_lba(&self) -> u32 {
+        self.size_in_lba
+    }
+}
+
+/// GPT partition entry.
+///
+/// UEFI specification: table 5-6.
+#[repr(C)]
+#[repr(packed)]
+#[derive(Clone, Copy, Debug)]
+pub struct GptPartitionEntry {
+    partition_type_guid: Guid,
```

Comment:
Should we define constants somewhere corresponding to `EFI System Partition` (`C12A7328-F81F-11D2-BA4B-00A0C93EC93B`) and `Partition containing a legacy MBR` (`024DEE41-33E7-11D3-9D69-0008C781F39F`)

Reply:
done, added a new `GptPartitionType` enum for this